### PR TITLE
Upgrades Xterm to v. 5

### LINKED ISSUE #13649: Upgrade XTerm to the latest
We should update xterm.js to the latest version: https://github.com/xtermjs/xterm.js/releases/tag/5.1.0

It has the benefit of being lighter and using the more efficient webgl renderer.

### PR DESCRIPTION
**Pull request by JasonWeill**
## References

Fixes #13649. Upgrades the Xterm dependency to version 5, modifies the code to make up for the removed `setOptions` function in it.

## Code changes

Updates Terminal package to use the new XtermJS 5.x API.
In [55cc717](https://github.com/jupyterlab/jupyterlab/pull/13685/commits/55cc717657e31e62e578df65cab298b5b100af17), WebGL addon is used if the browser supports it; otherwise fallback to canvas addon (as prior to this PR).
The asynchronous loading has changed as playing with import of file within our package may easily breaks what was done. So I [Frederic] went for a pattern used in more places where we asynchronously import the external library; this induced the need to buffer messages from the session waiting for the terminal to be loaded.

> If neither WebGL nor canvas addon is used, the terminal is rendered as html tags. This is to be avoided for performance reason.

## User-facing changes

None.

## Backwards-incompatible changes

None.

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks @JasonWeill.

Let's trigger a snapshot update since the UI tests seem to be failing.

**Comment by jtpio** [comment]
bot please update snapshots

**Comment by fcollonval** [comment]
@JasonWeill would you mind rebasing this PR taking into account the new xterm addon added in https://github.com/jupyterlab/jupyterlab/pull/13645

**Comment by JasonWeill** [comment]
@fcollonval Rebased, rebuilt, and updated this branch.

**Comment by fcollonval** [comment]
bot please update snapshots

**Review by krassowski — DISMISSED** [review]
Looks great! It seems that there is a regression related to selecting text in terminal. When you select anything the selection is not visible in the light theme (and barely visible in dark theme, though it might have been this way before?). It might be related to the rename of `ITheme.selection` to `ITheme.selectionBackground` in xterm.js 5.0

**Comment by krassowski** [comment]
I think that we need to change this interface and all usages:

https://github.com/jupyterlab/jupyterlab/blob/d4688fe784c2b7d8cb9af8a153af88c91fd8dcc8/packages/terminal/src/tokens.ts#L163-L170

**Comment by JasonWeill** [comment]
> Looks great! It seems that there is a regression related to selecting text in terminal. When you select anything the selection is not visible in the light theme (and barely visible in dark theme, though it might have been this way before?). It might be related to the rename of `ITheme.selection` to `ITheme.selectionBackground` in xterm.js 5.0

@krassowski I just updated the code and docs; the selections now have high contrast again

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by krassowski** [comment]
Looking again at the [changelog for 5.0](https://github.com/xtermjs/xterm.js/releases/tag/5.0.0), I wanted to highlight two things:
- there is now also `ITheme.selectionInactiveBackground`, should we add it too?
- (just for awareness) canvas is no longer the default renderer (and required code is not even installed by default):
> The canvas renderer has moved into an addon, significantly reducing the bundle size of the xterm module. This used to be the default renderer and is now recommended only as a fallback to the webgl addon if that does not work for some reason (eg. no webgl2 support).

**Comment by fcollonval** [comment]
Kicking the CI

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by fcollonval** [comment]
bot please update snapshots

File: packages/terminal/src/widget.ts [hunk 1]
```diff
@@ -61,66 +68,65 @@ export class Terminal extends Widget implements ITerminal.ITerminal {
...
       });
-    }
+  }
 
-    // Only run this initial connection logic once.
-    this.session.connectionStatusChanged.disconnect(
-      this._initialConnection,
-      this
-    );
+  get ready(): Promise<void> {
```
Comment (fcollonval) [inline packages/terminal/src/widget.ts:128]:
```suggestion
  /**
   * A promise that is fulfilled when the terminal is ready.
   */
  get ready(): Promise<void> {
```

**Review by fcollonval — APPROVED** [review]
Thanks @JasonWeill

**Comment by JasonWeill** [comment]
>     * there is now also `ITheme.selectionInactiveBackground`, should we add it too?

This is now added.

>     * (just for awareness) canvas is no longer the default renderer (and required code is not even installed by default):

The logic in the `terminal` package seems to favor the WebGL renderer if it can be used; do we need to do anything else as part of the upgrade?

**Comment by fcollonval** [comment]
> The logic in the `terminal` package seems to favor the WebGL renderer if it can be used; do we need to do anything else as part of the upgrade?

Yes, I did it in my rebase. Starting with xterm.js 5, the default renderer is to produce a real HTML tree. This is really inefficient (for what I saw every character is wrap in a span tag). So we need to load one of the renderer by default. The code I pushed is loading the WebGL renderer by default as it is favor over Canvas. But canvas is still used in case the browser does not support WebGL.

**Comment by fcollonval** [comment]
Thanks @JasonWeill  - the CI failures are not related.

**Pull request merged**